components/feature/downloads/src/main/java/mozilla/components/feature/downloads/share/ShareDownloadFeature.kt

```diff
@@ -0,0 +1,209 @@
+
+    @VisibleForTesting
+    internal fun startSharing(internetResource: ShareInternetResourceState) {
+        val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+            when (throwable) {
+                is InterruptedException -> {
+                    logger.warn("Share failed: operation timeout reached")
+                }
+                is IOException,
+                is RuntimeException,
+                is NullPointerException -> {
+                    logger.warn("Share failed: $throwable")
+                }
+            }
+        }
+
+        scope?.launch(coroutineExceptionHandler) {
+            withTimeout(OPERATION_TIMEOUT_MILLIS) {
+                val download = download(internetResource)
+                share(
+                    contentType = internetResource.contentType,
+                    filePath = download.canonicalPath,
+                    message = internetResource.url
+                )
+            }
+        }
+    }
+
+    @WorkerThread
+    @VisibleForTesting
+    internal fun download(internetResource: ShareInternetResourceState): File {
+        val request = Request(internetResource.url.sanitizeURL(), private = internetResource.private)
+        val response = if (internetResource.response == null) {
+            httpClient.fetch(request)
+        } else {
+            requireNotNull(internetResource.response)
+        }
+
+        if (response.status != Response.SUCCESS) {
+            // We experienced a problem trying to fetch the file, nothing more we can do.
+            throw(RuntimeException("Resource is not available to download"))
+        }
+
+        val fileExtension = '.' + getFileExtension(request.url)
+        val tempFile = getTempFile(fileExtension)
+        response.body.useStream { input ->
+            FileOutputStream(tempFile).use { output -> input.copyTo(output) }
+        }
+
+        return tempFile
+    }
+
+    @VisibleForTesting
+    internal fun share(
+        filePath: String,
+        contentType: String?,
+        subject: String? = null,
+        message: String? = null
+    ) = context.shareMedia(filePath, contentType, subject, message)
+
+    @VisibleForTesting
+    internal fun getFilename(fileExtension: String) =
+        Random.nextInt().absoluteValue.toString() + fileExtension
+
+    @VisibleForTesting
+    internal fun getTempFile(fileExtension: String) =
+        File(getMediaShareCacheDirectory(), getFilename(fileExtension))
+
+    @VisibleForTesting
+    internal fun getCacheDirectory() = File(context.cacheDir, cacherDirName)
+
+    @VisibleForTesting
+    internal fun getFileExtension(resourceUrl: String): String {
+        return if (resourceUrl.startsWith("data")) {
+            ("data:image\\/([a-zA-Z0-9-.+]+).*").toRegex()
+                .find(resourceUrl)?.groups?.get(1)?.value ?: DEFAULT_IMAGE_EXTENSION
+        } else {
+            MimeTypeMap.getFileExtensionFromUrl(resourceUrl).ifBlank { DEFAULT_IMAGE_EXTENSION }
```

Comment:
We have utility that for getting the name from `url` `DownloadUtils.guessFileName`

Reply:
Looked into this.
In the case of `https://mozilla.org/image.jpg` `DownloadUtils.guessFileName` would return `image.jpg` but my `getFileExtension` method is expected to return `jpg` in this case.
I don't see any advantage in using the `DownloadUtils` method and would prefer to keep the current implementation.

Reply:
Sorry, my mistake your are correct, I was a bit worry about partially duplicating some of the logic that we have on `DownloadUtils.guessFileName`, but your use case it is a bit different as you want the name from the URL as you want to create a temporal file name